code/dirs.py

```diff
@@ -260,3 +307,230 @@ def makedirs_if_not_exist(directory):
     except OSError as e:
         if e.errno != errno.EEXIST:
             raise
+
+def get_csidl(app_dir):
+    if app_dir == "local":
+        return 28 # CSIDL_LOCAL_APPDATA roaming app data
+    if app_dir == "roaming":
+        return 26 # CSIDL_APPDATA non-roaming app data
+    if app_dir == "home":
+        return 26 # CSIDL_PERSONAL non-roaming app data
+        
+class GUID(ctypes.Structure):   # [1]
+    _fields_ = [
+        ("Data1", wintypes.DWORD),
+        ("Data2", wintypes.WORD),
+        ("Data3", wintypes.WORD),
+        ("Data4", wintypes.BYTE * 8)
+    ] 
+
+    def __init__(self, l, w1, w2, b):
+        self.Data1 = l
+        self.Data2 = w1
+        self.Data3 = w2
+        self.Data4[:] = tuple((b & (0xff << pos * 8)) >> pos * 8 for pos in range(7, -1, -1))
+        
+
+def get_folder_id(app_dir):
+    if app_dir == "local":
+        return GUID(0xF1B32785, 0x6FBA, 0x4FCF, 0x9D557B8E7F157091) # non-roaming app data
+    if app_dir == "roaming":
+        return GUID(0x3EB685DB, 0x65F9, 0x4CF6, 0xA03AE3EF65729F3D) # roaming app data
+    if app_dir == "home":
+        return GUID(0x5E6C858F, 0x0E22, 0x4760, 0x9AFEEA3317B67173) # profile dir
+
+shell32_function = (
+    'SHGetKnownFolderPath'
+    'SHGetFolderPathW'
+    'SHGetFolderPathA'
+    'SHGetSpecialFolderPathW'
+    'SHGetSpecialFolderPathA'
+)
+
+def shell32_SHGetKnownFolderPath(func, app_dir):
+    buf = ctypes.c_wchar_p()
+    if func(ctypes.byref(ctypes.byref(get_folder_id(app_dir))), 0, 0, ctypes.byref(buf)):
+        return buf.value
+    return None
+
+def shell32_SHGetFolderPathW(func, app_dir):
+    buf = ctypes.create_unicode_buffer(ctypes.wintypes.MAX_PATH + 1)
+    if func(0, get_csidl(app_dir), 0 """SHGFP_TYPE_CURRENT""", buf):
+        return buf.value
+    return None
+
+def shell32_SHGetFolderPathA(func, app_dir):
+    buf = ctypes.create_string_buffer(ctypes.wintypes.MAX_PATH + 1)
+    if func(0, get_csidl(app_dir), 0 """SHGFP_TYPE_CURRENT""", buf)):
+        return buf.value
+    return None
+
+def shell32_SHGetSpecialFolderPathW(func, app_dir):
+    buf = ctypes.create_unicode_buffer(ctypes.wintypes.MAX_PATH + 1)
+    if func(0, buf, get_csidl(app_dir), True):
+        return buf.value
+    return None
+
+def shell32_SHGetSpecialFolderPathA(func, app_dir):
+    buf = ctypes.create_string_buffer(ctypes.wintypes.MAX_PATH + 1)
+    if (func(0, buf, get_csidl(app_dir), True)):
+        return buf.value
+    return None
+
+def get_win_folders():
+
+    # First try to import ctypes to call c win32 function.
+    # win32 function is the correct way to get win folder.
+    # but python is not shipped on windows with pywin32
+    # Let's do the hard way.
+    try:
+        import ctypes
+        import ctypes.wintypes
+    except:
```

Comment:
Please use `except ImportError:`.  We should definitely not catch every throwable here (we do not want to handle requests to shutdown or out or memory errors).